### PR TITLE
[ci] Add Vulkan NVIDIA benchmarks for clip and unet models

### LINKED ISSUE #12871: Add support for running Vulkan benchmarks on NVIDIA GPUs
### Request description

We are benchmarking Vulkan on mobile devices only. We should extend that to run on desktop for Stable Diffusion use cases.

### What component(s) does this issue relate to?

Other

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by antiagainst**
These models are parts of the stable diffusion model. This commit
enables compiling them with SIMT and tensorcore flavors and runs
on cloud NVIDIA GPUs.

Progress towards https://github.com/openxla/iree/issues/12635
Fixes https://github.com/openxla/iree/issues/12871

### CODE REVIEW

**Comment by powderluv** [comment]
Can't thank you enough for this one.

File: build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+      target_abi=iree_definitions.TargetABI.VULKAN_LINUX)
+
+  COMPILE_EXTRA_FLAGS = [
+      "--iree-stream-resource-index-bits=64", "--iree-vm-target-index-bits=64",
+      "--iree-preprocessing-pass-pipeline=builtin.module(func.func("
+      "iree-flow-detach-elementwise-from-named-ops,"
+      "iree-preprocessing-convert-conv2d-to-img2col,"
+      "iree-flow-convert-1x1-filter-conv2d-to-matmul,"
+      "iree-preprocessing-pad-linalg-ops{pad-size=32}))"
+  ]
```
Comment (GMNGeoffrey) [inline build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py]: Having implicit string concatenation across lines and list elements across lines is pretty confusing, IMO. Especially when the line elements have a comma inside the quote. Can we wrap this in parens, add an explicit `+` or somehow make this clearer

```suggestion
    COMPILE_EXTRA_FLAGS = [
        "--iree-stream-resource-index-bits=64",
        "--iree-vm-target-index-bits=64",
        ("--iree-preprocessing-pass-pipeline=builtin.module(func.func("
         "iree-flow-detach-elementwise-from-named-ops,"
         "iree-preprocessing-convert-conv2d-to-img2col,"
         "iree-flow-convert-1x1-filter-conv2d-to-matmul,"
         "iree-preprocessing-pad-linalg-ops{pad-size=32}))"),
    ]
```

File: build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+      target_backend=iree_definitions.TargetBackend.VULKAN_SPIRV,
+      target_abi=iree_definitions.TargetABI.VULKAN_LINUX)
+  PASCAL_TARGET = iree_definitions.CompileTarget(
+      target_architecture=common_definitions.DeviceArchitecture.NVIDIA_PASCAL,
+      target_backend=iree_definitions.TargetBackend.VULKAN_SPIRV,
+      target_abi=iree_definitions.TargetABI.VULKAN_LINUX)
+
+  COMPILE_EXTRA_FLAGS = [
+      "--iree-stream-resource-index-bits=64", "--iree-vm-target-index-bits=64",
+      "--iree-preprocessing-pass-pipeline=builtin.module(func.func("
```
Comment (GMNGeoffrey) [inline build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py]: This is a lot of custom stuff to get this to compile. Generally we want `iree-compile` to just work. Why is all this necessary? Should these configurations be tagged with "experimental flags"?

@MaheshRavishankar since you've had opinions about this sort of thing in the past.

File: build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+             List[iree_definitions.E2EModelRunConfig]]:
+    gen_configs = [
+        iree_definitions.ModuleGenerationConfig.build(
+            compile_config=compile_config,
+            imported_model=iree_definitions.ImportedModel.from_model(model))
+        for model in models
+    ]
+    sm80_devices = device_collections.DEFAULT_DEVICE_COLLECTION.query_device_specs(
+        # We use the same cloud NVIDIA GPU for Vulkan and CUDA.
+        architecture=common_definitions.DeviceArchitecture.CUDA_SM80,
```
Comment (pzread) [inline build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py]: Do we actually want to run benchmarks compiled for Pascal on A100, which is Ampere? Should we create the V100 machines to benchmark Pascal targets?

File: build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+      run_tags: Sequence[str] = [],
+  ) -> Tuple[List[iree_definitions.ModuleGenerationConfig],
+             List[iree_definitions.E2EModelRunConfig]]:
+    gen_configs = [
+        iree_definitions.ModuleGenerationConfig.build(
+            compile_config=compile_config,
+            imported_model=iree_definitions.ImportedModel.from_model(model))
+        for model in models
+    ]
+    sm80_devices = device_collections.DEFAULT_DEVICE_COLLECTION.query_device_specs(
```
Comment (pzread) [inline build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py]: I think we can update the A100 device spec to use DeviceArchitecture.NVIDIA_AMPERE

https://github.com/openxla/iree/blob/63fee4852bfa89e9b3bd84514092950ec9951a37/build_tools/python/e2e_test_framework/device_specs/gcp_specs.py#L20-L25

 And update the CUDA benchmarks to select the right device

https://github.com/openxla/iree/blob/63fee4852bfa89e9b3bd84514092950ec9951a37/build_tools/python/benchmark_suites/iree/cuda_benchmarks.py#L58-L60

File: build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py (hunk 3 above)
Comment (MaheshRavishankar) [reply build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py]: If they are compiled for Pascal, running the on V100 or A100 makes little difference

File: build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py (hunk 4 above)
Comment (antiagainst) [reply build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py]: Good points. Updated.

File: build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py (hunk 1 above)
Comment (antiagainst) [reply build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py]: Good points. Changed to implementation to use two lists, which should be more readable now!

File: build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py (hunk 2 above)
Comment (antiagainst) [reply build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py]: I think there is a difference here. This configuration is not something we are _experimenting_ a new compiler feature or whatever, it's actually the way we compile for _production_ for stable diffusion via nod.ai distribution channels. I'd say it's a different level so worth having special considerations w.r.t. support and guarantees.  I agree with the general direction of making the compiler just works so we don't need various flags, but it's still afar to reach that goal. While in the meanwhile, in the past few weeks I have been fixing quite a few regressions regarding this already, including https://github.com/openxla/iree/issues/13051, https://github.com/openxla/iree/issues/12670, https://github.com/openxla/iree/issues/12634, etc. Having CI coverage would avoid such continuous breakages for downstream deployment and avoid always spending a long time trying to triage and fix bugs afterwards. :)

File: build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py (hunk 3 above)
Comment (antiagainst) [reply build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py]: +1. Ampere devices can certainly support binaries generated for earlier generations. We don't care about deployment on V100 (or even A100 for now). This is more making sure that the compiled binaries runs fine.

File: build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+  ) -> Tuple[List[iree_definitions.ModuleGenerationConfig],
+             List[iree_definitions.E2EModelRunConfig]]:
+    gen_configs = [
+        iree_definitions.ModuleGenerationConfig.build(
+            compile_config=compile_config,
+            imported_model=iree_definitions.ImportedModel.from_model(model))
+        for model in models
+    ]
+    ampere_devices = device_collections.DEFAULT_DEVICE_COLLECTION.query_device_specs(
+        # We use the same cloud NVIDIA GPU for Vulkan and CUDA.
```
Comment (pzread) [inline build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py]: This comment seems to be unrelated now.

File: build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+      run_tags: Sequence[str] = [],
+  ) -> Tuple[List[iree_definitions.ModuleGenerationConfig],
+             List[iree_definitions.E2EModelRunConfig]]:
+    gen_configs = [
+        iree_definitions.ModuleGenerationConfig.build(
+            compile_config=compile_config,
+            imported_model=iree_definitions.ImportedModel.from_model(model))
+        for model in models
+    ]
+    ampere_devices = device_collections.DEFAULT_DEVICE_COLLECTION.query_device_specs(
```
Comment (pzread) [inline build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py:75]: Can you add a comment to explain why we benchmarks pascal and ampere targets on the ampere GPU?

**Review by pzread — APPROVED** [review]
LGTM. But we might want to resolve the concern on non-default/specialized compile flags before merge.

File: build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py (hunk 6 above)
Comment (antiagainst) [reply build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py:75]: Yes, done now.

File: build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py (hunk 2 above)
Comment (antiagainst) [reply build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py]: Ping on this to see if there are further comments. I think I can add `production` in the tags if that makes things clearer.

Comment (GMNGeoffrey) [reply build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py]: If we're not able to turn them on by default then they seems pretty experimental, regardless of where people are using them. I think "production" would communicate the wrong thing for the same reason. I would still prefer to tag these as "experimental flags", since that's consistent with all our other custom flag configs, but if you want to call them "custom flags" or something instead, that's fine.

Comment (antiagainst) [reply build_tools/python/benchmark_suites/iree/vulkan_nvidia_benchmarks.py]: SG. Added "experimental-flags" label.
